### ISSUE TITLE
Hiking Path Difficulty Ratings Based on Route Plotted

### ISSUE THREAD

**Issue by DeathfireD** [organicmaps/organicmaps#9525]
OpenStreetMap's rating system for hikes is pretty useless here in the U.S. We don't have a standard ranking system and because of that most of the OpenStreetMap hiking trails here don't have any consistent ratings on them. When looking at the hikes, there's not really a good way to tell how difficult a hike is which then forces me to open AllTrails, HikingProject or some other app that uses a proprietary algorithm to display a trail's difficulty. They all seem to be pretty consistent with their ratings though.

### How it should work
My proposal is to have an option that can be turned on in the settings for hiking/walking. When ON, Organic Maps would calculate a route's difficulty based on the information it sees between the starting point and the end point that you are planning to navigating to. The typical blue path line would be changed to green for easy, blue for moderate, and red for hard / Strenuous (colors subject to change depending on theme used). When the start button is pressed to begin navigating, the blue line would also change to whatever the difficulty color is and re-adjust if the user "adds a stop" that forces the path to be recalculated.

Distance, Terrain, and Elevation gained should be used for this rating. 

Here's a link to a good starting point. You could always adjust it over time or possibly even allow users to choose from different difficulty options. https://www.smcgov.org/parks/county-hiking-trail-difficulty-rating-system
This site also has a lot of really good equations you could use - https://trailsnh.com/tools/hiking-difficulty.php

Easy:
    Up to two miles long
    Flat or nearly level terrain
    Little change in elevation (< 250’)

Moderate:
    Two to four miles long
    Moderate terrain
    Some change in elevation (250’ to 750’)

Hard / Strenuous:
    More than four miles long
    Variable terrain
    More change in elevation (> 750’)

### Links to resources for different ways to calculate difficulty:
https://www.smcgov.org/parks/county-hiking-trail-difficulty-rating-system
https://trailsnh.com/tools/hiking-difficulty.php
https://www.fs.usda.gov/detailfull/willamette/recreation/hiking/?cid=stelprdb5103006
https://www.hikingproject.com/help/21/overview-of-site-name-features

Another suggestion would be to include this difficulty rating AND the gradient into the Navigation section at the bottom. See this feature proposal and add difficulty to it.
https://github.com/organicmaps/organicmaps/issues/9200

Thanks!

**Comment by DeathfireD** [comment]
After digging through the current pull requests of features people are working on and other proposals, I feel it would be best to instead add a small icon to indicate a Trail/Path's difficulty instead of making the line you follow change colors. I've mocked up an example of what I mean.

The left photo shows trail's rating between the miles and elevation gain/lost. The background would change colors based on the difficulty calculated. Green being easy, Orange being Moderate, Red being Hard. I've also added AVG Incline percent to the gradient as it'd be a good addition to the information for hikers.

The right photo shows what it will look like after you press the start button. When you swipe up from the bottom you will see the trail's difficulty rating as well as the gradient showing the current incline that you are at. This was proposed on this ticket https://github.com/organicmaps/organicmaps/issues/9200 but I added it in because I feel it'd be a good addition of information for hikers.

![Trail-difficulty-Proposal](https://github.com/user-attachments/assets/b0b04e06-fbfb-4f46-9447-df5088a04104)


**Formula changes**
 After thinking about it more, the original formula could be problematic. Miles for a trail might be long, say 4 but elevation gain might be low 150' and terrain/slop might be flat. As you can see, based off the original formula, the long miles would not make it an easy trail but the elevation and terrain would. How would this get calculated? Is it easy because 2 criteria beat 1?

So I came up with a point system.
**Length |**
```
< 2 miles | 1 point
2-4 miles | 2 points
>4-6 miles | 3 points
>6-8 miles | 4 points
> 8 miles | 5 points
```

**Elevation Gain |**  
```
< 250 feet | 1 point
250-750 feet | 2 points
>750-1500 feet | 3 points
>1500-2500 feet | 4 points
> 2500 feet | 5 points
```

**AVG Terrain/Slope |**  
```
Flat | 1 point
Moderate | 2 points
Steep or technical | 3 points
```

**Difficulty Ratings:**
```
Easy: Total points 3-6
Moderate: Total points 7-9
Hard: Total points 10+
```

So in the example I gave before, 4 miles would net you 2 points, 150' elevation gain would net you 1 point and flat terrain would net you 1 point. Total points would be 4 which is considered an easy trail.

Another example: Mount Washington in NH. 8.6m with 5190' elevation gain and terrain is steep and technical. 
length = 5 points, elevation = 5 points, difficulty = 3 points so 13 total points which makes it Hard.

I've used this point system to compare against the app All trails and similar apps and it seems to be accurate with their trail ratings. The only thing is avg terrain/slope/incline is not given for hikes so I've been choosing whether I think it should be flat, moderate, or steep/technical. So I don't know how the rating will change when Organic maps uses the true value when calculated. It may nee some adjustments but aside from that, I think this is a good start or something like it.
Look on this site, specifically the steepness chart and the formulas used to get the grade and angle. They may be useful. 
https://trailsnh.com/tools/hiking-difficulty.php

**Comment by DeathfireD** [comment]
To add to this, it would also be nice if when you click on the Trail Difficulty colored button, it would open a detailed breakdown of the points so if the rating doesn't look right, people can screenshot the detail page and include it on their github report to help determine if the formula needs adjusting. Here's a mock up.
![difficulty-proposal-details2](https://github.com/user-attachments/assets/ce26c1f4-fca9-4796-9cf6-fdfefbf924a3)

**Comment by pastk** [comment]
Its an interesting idea.

Perhaps the rating should include other relevant tags/info like sac_scale, surface, smoothness, trail_visibility...
Also check the existing visual grading:
- https://github.com/organicmaps/organicmaps/pull/7926

OR it should be obvious the rating is limited to the incline.

Probably the biggest challenge here is to develop good heuristics to make such ratings precise enough to be useful.

**Comment by DeathfireD** [comment]
@pastk  I agree that incorporating data points like sac_scale, trail_visibility, surface, trailblazed...etc would be nice, however I do think we would face major consistency challenges when trying to calculate the difficulty level using them. For instance, sac_scale can be wildly different depending on where you live in the world. It's not standardized in the U.S and probably many other locations. Many trails in my area are incorrectly categorized as "hiking" or "mountain_hiking," making it difficult to distinguish a trail's true difficulty level. A hike may have a sac_scale rating of "mountain_hiking" but on the popular hiking apps (Alltrails and Mountain Project), it's got a rating of Easy because it's basically a flat walk in the woods. A very hard hike that requires crampons and other advanced equipment may also have a sac_scale of "mountain_hiking" but in the hiking apps, it has a rating of Hard. It feels almost like people use it to label the sport instead of following the guidelines on the OpenStreetMap Wiki. We can't reliably use this to factor in a trail's difficulty.

The other data points I could see being useful however not every trail has them, which again, I think would mess with the consistency of the ratings. I think instead of trying to add them in to the equation, they should be displayed on a "Trail Report" type page that would basically summarize all the important information for the plotted path you plan to take.

ex.
Trail Name 1 | Segment Length: 1 mile | Avg Slope: 15% | Elevation Gain: 400' | Sac Scale: Mountain Hiking | Surface: Ground | Trail Visibility: Good | Trailblazed: Yes, Poles

Trail Name 2 | Segment Length: 1.4 miles | Avg Slope: 25% | Elevation Gain: 800' | Sac Scale: Demanding Mountain Hiking | Surface: Ground | Trail Visibility: Bad | Trailblazed: Yes, Cairns

Trail Name 3 | Segment Length: .4 miles | Avg Slope: 5% | Elevation Gain: 85' | Surface: Ground | 

I think for now it would be better to try and keep the calculations as simple and reliable as possible. Use the plotted path's length, Elevation Gain, Avg Slope, and possibly Avg Angle. From my own testing, my point system has been relatively correct for U.S trails vs what's in the Hiking Apps. I'm still trying to figure out the Slope and Angle points at the moment though. What I have in my screenshot is incorrect.

**Comment by pastk** [comment]
> Many trails in my area are incorrectly categorized as "hiking" or "mountain_hiking," making it difficult to distinguish a trail's true difficulty level. 

One way to deal with it is to consider sac_scale in a rough way, e.g. "hiking" and "mountain_hiking" could fall into the same "easy" category and the rating would be mostly determined by incline, distance, etc.
E.g. in #7926 we start to display the trail dotted (more difficult) only if its higher than T3 "demanding_mountain_hiking".
So its a matter of processing the data and applying weights that make sense.

> The other data points I could see being useful however not every trail has them, which again, I think would mess with the consistency of the ratings. I think instead of trying to add them in to the equation, they should be displayed on a "Trail Report" type page that would basically summarize all the important information for the plotted path you plan to take.
> 

For sure:
- https://github.com/organicmaps/organicmaps/issues/4898

(maybe add your examples there?)

**Comment by biodranik** [comment]
From our user:
>I would probably draw inspiration from the graphics of the map service of the Austrian web platform Bergfex.at (alternativelly TracesTrack Topo layer in osm.org). In attachment, there is an example from Austria/"Kaiser Gebirge" Mountains/ Peak "Treffauer"..The trails here have a difficulty from T2 to T6. In this regard, please see the legend as well.
<img width="1525" height="1078" alt="Image" src="https://github.com/user-attachments/assets/76fdf289-6150-4381-a9c8-8998fd9c1cf4" />

**Comment by DeathfireD** [comment]
> From our user:
> 
> > I would probably draw inspiration from the graphics of the map service of the Austrian web platform Bergfex.at (alternativelly TracesTrack Topo layer in osm.org). In attachment, there is an example from Austria/"Kaiser Gebirge" Mountains/ Peak "Treffauer"..The trails here have a difficulty from T2 to T6. In this regard, please see the legend as well.
> 

@biodranik Wasn't sac scale dots and dashes already added or am I misunderstanding your post? https://github.com/organicmaps/organicmaps/pull/7810

My proposal is different. It's to dynamically determine the trail's difficulty by using the total distance, elevation gained, and average slope from the "Route from" to the "Route to" points that a user places or from their current GPS location to their "Route to" location. Display that difficulty level in a colored badge (seen above) with the actual word in it (for color blind people). 

Sac Scale is not a helpful indicator of trail difficulty for most countries, the Americas mainly. We don't follow it and if we did, most of our trails would just be T2 and T3. This is why I'm suggesting we use a formula to determine the difficulty for just the Hiking Layer and allow users the ability to turn it on or off.

**Comment by biodranik** [comment]
Sorry, the comment above was about a different issue.

The idea to calculate the difficulty dynamically (ideally taking into the account user's preferences/profile/abilities/local device's history) and show it is good, if users can easily estimate the route in such a way (and understand it properly).

Would it be possible to also calculate/show the difficulty in the local region's terms, that users may already know?

**Comment by DeathfireD** [comment]
@biodranik I'm not sure what you mean by calculate/show in the local region's terms?

If you're referring to how a trail's difficulty is visually shown...it's pretty well established that:

**Easy:**
Color: Green
Word: Easy
Shape: Circle

**Medium:**
Color: Blue
Word: Medium / Moderate
Shape: Square

**Hard:**
Color: Black
Word: Hard / Difficult
Shape: Diamond

and in some cases

**Extremely difficult:**
Color: Black
Word: Extremely Difficult
Shape: Double black diamonds

But maybe you give the user the ability to customize this in settings?

**Comment by biodranik** [comment]
Is it established in the same way _all over the world_?

**Comment by DeathfireD** [comment]
No, I can't say it's established for _all over the world_  like skiing ratings are. That's kind of the problem. Some rating systems (eg. Sac scale, Aussie Scale...etc) have more then 3 levels of difficulty but can generally be combined and sorted into Easy, Moderate, Hard though.

Apps like AllTrails, Komoot, and Hiking Project work world wide and the rating systems are colored and worded similarly so any layman can understand. AllTrails uses 3 difficult levels (Easy, Moderate, Hard) generated using their own proprietary calculations that seem to be based on distance, elevation, slope. Komoot uses 3 (Easy, Moderate, Hard) which uses a combination of distance, estimated time, elevation, slope, sac scale. Hiking Project uses 6 (Easy, Easy-Intermediate, Intermediate, Intermediate-Difficult, Difficult, Very Difficult). It bases it's ratings off of user submitted ratings of a trail, using the rating with the most votes as the trail's difficulty rating.

In a perfect world, Organic Maps would give the user the ability to tweak and choose the calculations and difficulty ratings in settings but I'm hopping we can at least get the ball rolling and have something basic so Hikers like myself don't have to keep juggling between multiple hiking apps to get a general idea of how hard a trail may be.

If it helps, I have an excel sheet that I've been working on that has a mixture of modified difficulty calculation formulas and how they would score a trail from a list of trails across the world. I display two Difficulty Ratings per trail. The first is the rating that was calculated the most by all the calculations and the second is the highest difficulty that was determined. I can share it after I clean it up a bit if you'd like.

**Comment by biodranik** [comment]
Thanks for the details! It would be interesting to develop a difficulty classification that is understandable to everyone.

Maybe calculate it for different routes and compare with other apps?

Easy, Moderate, Hard look okay, but it feels like Hard can be very different (e.g., requiring equipment or not). Maybe a 4th grade of something like "Professional", to alert users about dangers?
